Improve performance of readuntil

omus opened the pull request:
Was looking into `readuntil` and found a way to improve performance. Below are some of the benchmarks I used for comparison:
```julia
julia> function readuntil_old(s::IO, r::AbstractString)
           l = length(r)
           if l == 0
               return ""
           end
           out = IOBuffer()
           m = Array{Char}(l)  # last part of stream to match
           t = collect(r)
           i = 0
           while !eof(s)
               i += 1
               c = read(s, Char)
               write(out, c)
               if i <= l
                   m[i] = c
               else
                   # shift to last part of s
                   for j = 2:l
                       m[j-1] = m[j]
                   end
                   m[l] = c
               end
               if i >= l && m == t
                   break
               end
           end
           return String(take!(out))
       end
readuntil_old (generic function with 1 method)

julia> function readuntil_new(s::IO, t::AbstractString)
           l = length(t)
           if l == 0
               return ""
           end
           out = IOBuffer()
           i = 1
           while !eof(s)
               c = read(s, Char)
               write(out, c)
               i = c == t[i] ? i + 1 : 1
               if i > l
                   break
               end
           end
           return String(take!(out))
       end

readuntil_new (generic function with 1 method)

julia> using BenchmarkTools

julia> str = String([rand('A':'Z', 50000); '1']);

julia> io = IOBuffer(str);

julia> goal = str[end-1000:end];
julia> b1 = @benchmark readuntil_old(seekstart($io), $goal);
julia> b2 = @benchmark readuntil_new(seekstart($io), $goal);
julia> judge(median(b2),median(b1))
BenchmarkTools.TrialJudgement: 
  time:   -95.01% => improvement (5.00% tolerance)
  memory: -10.88% => improvement (1.00% tolerance)

julia> goal = str[end:end];
julia> b1 = @benchmark readuntil_old(seekstart($io), $goal);
julia> b2 = @benchmark readuntil_new(seekstart($io), $goal);
julia> judge(median(b2),median(b1))
BenchmarkTools.TrialJudgement: 
  time:   +2.53% => invariant (5.00% tolerance)
  memory: -0.28% => invariant (1.00% tolerance)

julia> goal = str;
julia> b1 = @benchmark readuntil_old($(seekstart(io)), $goal);
julia> b2 = @benchmark readuntil_new($(seekstart(io)), $goal);
julia>judge(median(b2),median(b1))
BenchmarkTools.TrialJudgement: 
  time:   +4.36% => invariant (5.00% tolerance)
  memory: -85.45% => improvement (1.00% tolerance)
```

I'll add these benchmarks to BaseBenchmarks.jl

omus committed:
Improve performance of readuntil using strings

vtjnash commented on base/io.jl:
```diff
@@ -445,33 +445,18 @@ function readuntil{T}(s::IO, delim::T)
-            m[l] = c
-        end
-        if i >= l && m == t
+        i = c == t[i] ? i + 1 : 1
```
Need to advance 'i' here by a whole character to handle Unicode. I think it'll be easiest to do that by using start/next/done instead of 1/+1/l

omus replied:
In the original code `i` represents a character number and not a index into the string.

vtjnash commented:
I think this proposed algorithm would fail with a terminator like "aab" matching against "aaab" (or in general, any string where the first character is repeated anywhere else in the string)

omus committed:
Add backtracking

omus commented:
Benchmarks for latest version. Note the original benchmarks weren't rewinding the stream correctly so I also update the original description.
```julia
julia> using BenchmarkTools
julia> str = String([rand('A':'Z', 50000); '1']);
julia> io = IOBuffer(str);

julia> goal = str[end-1000:end];
julia> b1 = @benchmark readuntil_old(seekstart($io), $goal);
julia> b2 = @benchmark readuntil_new(seekstart($io), $goal);
julia>judge(median(b2),median(b1))
BenchmarkTools.TrialJudgement: 
  time:   -95.66% => improvement (5.00% tolerance)
  memory: +5.27% => regression (1.00% tolerance)

julia> goal = str[end:end];
julia> b1 = @benchmark readuntil_old(seekstart($io), $goal);
julia> b2 = @benchmark readuntil_new(seekstart($io), $goal);
julia>judge(median(b2),median(b1))
BenchmarkTools.TrialJudgement: 
  time:   -16.59% => improvement (5.00% tolerance)
  memory: +0.00% => invariant (1.00% tolerance)

julia> goal = str;
julia> b1 = @benchmark readuntil_old(seekstart($io), $goal);
julia> b2 = @benchmark readuntil_new(seekstart($io), $goal);
julia>judge(median(b2),median(b1))
BenchmarkTools.TrialJudgement: 
  time:   +0.27% => invariant (5.00% tolerance)
  memory: +42.69% => regression (1.00% tolerance)
```

vtjnash committed:
Improve performance of readuntil using strings

Heavily inspired by omus and #20621

omus commented:
Travis failure appears unrelated.

omus committed:
Revise test to be completely unoffensive

Looking over my original test I realized it potentially could be
offensive. I've used a different example to avoid any potential issues.

omus committed:
readuntil with on-the-fly backtrack caching

Caches backtracking information as it is needed. Using a SparseVector
which has a lower memory footprint than Vector but is more performant
than Dict.

omus commented:
`O(n + k)` version of the algorithm. I ended up storing the backtracking information in a SparseVector which is more memory efficient than an Vector and computationally faster than a Dict.

omus commented:
I have a unicode test I want to add but currently that results in:
```julia
Error During Test
  Test threw an exception of type Base.UVError
  Expression: readuntil(io(t), s) == m
  read: network is down (ENETDOWN)
```

omus committed:
Add unicode test for readuntil

Skip testing the I/O producers "File" and "PipeEndpoint" when working
with unicode.

omus committed:
Remove need for readuntil file

StefanKarpinski commented:
Wat?

omus commented:
It looks like the I/O producers "File" and "PipeEndpoint" choke with unicode input in test/read.jl

tkelman commented:
https://github.com/JuliaArchive/LegacyStrings.jl/issues/4

vtjnash committed:
[wip] reduce code duplication, allow generalized Int indexes

vtjnash committed:
refactor into single method instead of a type,
makes it possible to use readuntil with any array (indexable) object
and optimizes a few more cases

vtjnash commented:
@omus PTAL at my latest updates to your code :)

omus commented:
Looks really solid. I'll try to dig up some of my old benchmarks and give this a spin.

vtjnash commented:
To be slightly more fair to the current algorithm, I was using the following modified version:
```julia
function readuntil_old(s::IO, r::Vector{UInt8})
                         l = sizeof(r)
                         if l == 0
                             return ""
                         end
                         out = Base.StringVector(0)
                         m = Array{UInt8}(l)  # last part of stream to match
                         i = 0
                         while !eof(s)
                             i += 1
                             c = read(s, UInt8)
                             push!(out, c)
                             if i <= l
                                 m[i] = c
                             else
                                 # shift to last part of s
                                 for j = 2:l
                                     m[j-1] = m[j]
                                 end
                                 m[l] = c
                             end
                             if i >= l && m == r
                                 break
                             end
                         end
                         return String(out)
                     end
 @time let g2 = Vector{UInt8}(goal); for i in 1:100; readuntil_old(seekstart(io), g2); end; end
```

I also added the worst case to my tests, where the new algo in this PR really shines:
```julia
str = ("A" ^ 50000) * "B";
goal = ("A" ^ 5000) * "Z";
```

omus commented:
I think you need to remove the `String(...)` from `readuntil_old` to make it a fair fight.

The only benchmark where the old algorithm was slightly faster is in this case:
```julia
goal = "A" ^ 50000;
str = "A" ^ 50000;
```

omus commented:
I'll make a PR to BaseBenchmarks.

ararslan commented:
Nanosoldier is now ready to run the `readuntil` benchmarks whenever you are.

omus commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/20621#issuecomment-331321260):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/1fb44196654c8776f09e829402a3bb2f19536f65/b797b9d_vs_ff9fb48/report.md)
cc @ararslan

omus commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/20621#issuecomment-331880613) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/2cfc902dbce129165d61051151f9bb686a38f3f0/ec5cc59_vs_ee7b0b8/report.md). cc @ararslan

vtjnash reviewed (approved):
please squash when merging

omus commented:
The new changes look great!
